server/app/jobs/deployment_cleanup_job.rb

```diff
@@ -0,0 +1,32 @@
+require_relative '../services/logging'
+
+class DeploymentCleanupJob
+  include Celluloid
+  include CurrentLeader
+  include Logging
+
+  def initialize(perform = true)
+    async.perform if perform
+  end
+
+  def perform
+    info 'starting to cleanup old deployments'
+    loop do
+      sleep 5.minute.to_i
+      if leader?
+        destroy_old_deployments
+      end
+    end
+  end
+
+  def destroy_old_deployments
+    GridService.each do |s|
+      return if s.grid_service_deploys.finished.count <= 100
+      info "cleaning old deployments for service #{s.to_path}"
+      # Find the last 100th deployments id
+      id = s.grid_service_deploys.finished.desc('_id').limit(100).to_a.last._id
+      # Now delete all deployments older than the 100th one
+      s.grid_service_deploys.finished.where(:_id.lt => id).delete
```

Comment:
Is there possibility to leak deploys if they never finish?

Reply:
There's the `deploying` => `{:started_at.gt => TIMEOUT.ago , finished_at: nil}` rule for detecting such dead deploys for the case of e.g. a server worker process crashing hard in the middle of a deploy.

If you also want to consider those, then the scope could be something like `scope :complete, { any_of({:finished_at.ne => nil}, {:started_at.lt => TIMEOUT.ago , finished_at: nil}) }`, I think.

Reply:
Initially I thought that it might make sense to leave those un-finished deploys to hang-around, it might give the ability to actually figure out what's making those "leak".